Microsoft.Research/ContractAdornments/CSharp.Roslyn/Intellisense/IntellisenseContractsHelper.cs

```diff
@@ -0,0 +1,395 @@
+      var nodeInQuestion = leafNode.Parent;
+      SyntaxNode ptn = null;
+
+      while (nodeInQuestion != null) {
+        //Is the node in question a node call?
+        var asCall = nodeInQuestion as InvocationExpressionSyntax;
+        if (asCall != null) {
+          ptn = asCall;
+          break;
+        }
+
+        var asCtorCall = nodeInQuestion as ObjectCreationExpressionSyntax;
+        if (asCtorCall != null) {
+          ptn = asCtorCall;
+          break;
+        }
+
+        //Climb higher up our ancestry for the next iteration
+        nodeInQuestion = nodeInQuestion.Parent;
+      }
+
+      //Did we successfully find a call node?
+      return ptn;
+    }
+    public static SyntaxNode GetTargetAtTriggerPoint(ITrackingPoint triggerPoint, ITextSnapshot snapshot, SyntaxTree parseTree)
+    {
+      Contract.Requires(snapshot != null);
+      Contract.Requires(parseTree != null);
+      Contract.Requires(triggerPoint != null);
+
+      SyntaxNode syntaxRoot;
+      if (!parseTree.TryGetRoot(out syntaxRoot))
+        return null;
+
+      var leafToken = syntaxRoot.FindToken(triggerPoint.GetPosition(snapshot), false);
+      if (leafToken.IsKind(SyntaxKind.None))
+        return null;
+
+      var leafNode = leafToken.Parent;
+
+      var asPropDecl = leafNode as PropertyDeclarationSyntax;
+      if (asPropDecl != null)
+      {
+          return asPropDecl;
+      }
+
+      //Is our leaf node a name?
+      var asName = leafNode as IdentifierNameSyntax;
+      if (asName == null)
+        return null;
+
+      //Is anyone in our ancestry a call node?
+      var nodeInQuestion = leafNode;
+      var lastNonBinaryOperatorNode = leafNode;
+      var lastNode = leafNode;
+      while (nodeInQuestion != null) {
+
+        //Is the node in question a node call?
+        var asCall = nodeInQuestion as InvocationExpressionSyntax;
+        if (asCall != null) {
+
+          //Make sure we aren't on the right side of the call
+          MemberAccessExpressionSyntax memberAccessExpression = asCall.Expression as MemberAccessExpressionSyntax;
+          if (memberAccessExpression == null || memberAccessExpression.Name == leafNode)
+            return null;
+
+          return asCall;
+        }
+
+        var asProp = nodeInQuestion as MemberAccessExpressionSyntax;
+        if (asProp != null)
+        {
+            //Make sure we aren't on the left side of the dot
+            if (asProp.Name == lastNode && asProp.Name is IdentifierNameSyntax)
+            {
+                return asProp;
+            }
+        }
+
+        var asCtorCall = nodeInQuestion as ObjectCreationExpressionSyntax;
+        if (asCtorCall != null)
+        {
+            return asCtorCall;
+        }
+
+        var declNode = nodeInQuestion as MemberDeclarationSyntax;
+        if (declNode != null)
+        {
+            if (lastNode == leafNode) { return declNode; }
+            return null;
+        }
+
+#warning What exactly is this doing?
```

Comment:
:memo: Someone will need to figure out what this was doing, but it didn't seem to be essential for the initial pull request. I am in favor of creating a new issue to resolve this warning after this pull request is merged.